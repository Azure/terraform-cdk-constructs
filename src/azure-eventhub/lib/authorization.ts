import {
  EventhubAuthorizationRule,
  EventhubAuthorizationRuleConfig,
} from "@cdktf/provider-azurerm/lib/eventhub-authorization-rule";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import { Vault } from "../../azure-keyvault";

export interface AuthorizationRuleProps {
  readonly name: string;
  /**
   * The name of the resource group in which the EventHub's parent Namespace exists.
   */
  readonly listen?: boolean;
  /**
   * Does this Authorization Rule have permissions to Send to the Event Hub?
   * @default false
   */
  readonly send?: boolean;
  /**
   * Does this Authorization Rule have permissions to Manage to the Event Hub? When this property is true - both listen and send must be too.
   * @default false
   */
  readonly manage?: boolean;
}

export class AuthorizationRule extends Construct {
  readonly ehInstanceAuthProps: EventhubAuthorizationRuleConfig;
  private readonly primaryConnectionString: string;
  private readonly primaryKey: string;

  /**
   * Constructs a new Authorization Rule for an Azure Event Hub.
   *
   * This class creates an authorization rule which defines the permissions granted to users and applications
   * for accessing and managing the Event Hub. An Authorization Rule can grant listening, sending, and full manage
   * permissions based on the properties specified.
   *
   * @param scope - The scope in which to define this construct, typically used for managing lifecycles and creation order.
   * @param name - The unique name for this instance of the Authorization Rule.
   * @param ehInstanceAuthProps - The properties for configuring the Authorization Rule. The properties include:
   *                - `name`: Required. The name of the Authorization Rule.
   *                - `listen`: Optional. Specifies whether the rule allows listening to the Event Hub. Defaults to false.
   *                - `send`: Optional. Specifies whether the rule allows sending events to the Event Hub. Defaults to false.
   *                - `manage`: Optional. Specifies whether the rule allows managing the Event Hub, including sending and listening. Defaults to false.
   *                When `manage` is true, both `listen` and `send` are implicitly set to true.
   *
   * Example usage:
   * ```typescript
   * const authRule = new AuthorizationRule(this, 'exampleAuthRule', {
   *   name: 'myAuthRule',
   *   listen: true,
   *   send: true,
   *   manage: false // Only listening and sending are enabled; not managing.
   * });
   * ```
   *
   * @remarks
   * The primary connection string and primary key are accessible after the instance creation,
   * allowing for integration with other Azure services or client applications.
   */
  constructor(
    scope: Construct,
    name: string,
    ehInstanceAuthProps: EventhubAuthorizationRuleConfig,
  ) {
    super(scope, name);

    this.ehInstanceAuthProps = ehInstanceAuthProps;

    // Fix input properties
    let listen = ehInstanceAuthProps.listen || false;
    let send = ehInstanceAuthProps.send || false;
    let manage = ehInstanceAuthProps.manage || false;

    if (manage) {
      listen = true;
      send = true;
    }

    const eventhubauthrule = new EventhubAuthorizationRule(
      this,
      ehInstanceAuthProps.name,
      {
        name: ehInstanceAuthProps.name,
        resourceGroupName: this.ehInstanceAuthProps.resourceGroupName,
        namespaceName: this.ehInstanceAuthProps.namespaceName,
        eventhubName: this.ehInstanceAuthProps.eventhubName,
        listen: listen,
        send: send,
        manage: manage,
      },
    );

    // Outputs
    this.primaryConnectionString = eventhubauthrule.primaryConnectionString;
    this.primaryKey = eventhubauthrule.primaryKey;

    const cdktfTerraformOutputEventhubAuthPrimaryConnectionString =
      new cdktf.TerraformOutput(this, "primary_connection_string", {
        value: eventhubauthrule.primaryConnectionString,
        sensitive: true,
      });
    const cdktfTerraformOutputEventhubAuthPrimaryKey =
      new cdktf.TerraformOutput(this, "primary_key", {
        value: eventhubauthrule.primaryKey,
        sensitive: true,
      });
    cdktfTerraformOutputEventhubAuthPrimaryConnectionString.overrideLogicalId(
      "primary_connection_string",
    );
    cdktfTerraformOutputEventhubAuthPrimaryKey.overrideLogicalId("primary_key");
  }

  addPrimaryConnectionStringToVault(
    vault: Vault,
    name: string,
    expirationDate?: string,
  ) {
    vault.addSecret(name, this.primaryConnectionString, expirationDate);
  }

  addPrimaryKeyToVault(vault: Vault, name: string, expirationDate?: string) {
    vault.addSecret(name, this.primaryKey, expirationDate);
  }
}
