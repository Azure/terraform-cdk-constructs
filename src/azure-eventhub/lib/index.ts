export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
