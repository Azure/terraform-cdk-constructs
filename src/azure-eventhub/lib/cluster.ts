import { EventhubCluster } from "@cdktf/provider-azurerm/lib/eventhub-cluster";
import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import { AzureResource } from "../../core-azure/lib";

export interface ClusterProps {
  /**
   * The name of the Resource Group in which to create the EventHub Cluster.
   */
  readonly resourceGroup: ResourceGroup;
  readonly name: string;
  /**
   * The SKU name of the EventHub Cluster. The only supported value at this time is Dedicated_1.
   * @default "Dedicated_1"
   */
  readonly skuName?: string;
  /**
   * The tags to assign to the Application Insights resource.
   */
  readonly tags?: { [key: string]: string };
}

export class Cluster extends AzureResource {
  readonly ehClusterProps: ClusterProps;
  public id: string;
  public resourceGroup: ResourceGroup;

  /**
   * Constructs a new Event Hub Cluster.
   *
   * This class creates an Azure Event Hub Cluster which is a dedicated capacity resource for handling
   * high-throughput, low-latency event ingestion and streaming. It is used in scenarios where you need
   * predictable performance and cost regardless of the volume of data ingress or number of downstream
   * event consumers.
   *
   * @param scope - The scope in which to define this construct, usually representing the Cloud Development Kit (CDK) stack.
   * @param name - The unique name for this instance of the Event Hub Cluster.
   * @param ehClusterProps - The properties for configuring the Event Hub Cluster. These properties include:
   *                - `resourceGroup`: Required. The Azure Resource Group in which the cluster will be deployed.
   *                - `name`: Required. The name of the Event Hub Cluster.
   *                - `skuName`: Optional. The SKU name for the cluster, which determines the pricing and capabilities.
   *                             Currently, the only supported value is "Dedicated_1". Defaults to "Dedicated_1" if not specified.
   *                - `tags`: Optional. Tags for resource management and categorization.
   *
   * Example usage:
   * ```typescript
   * const eventHubCluster = new Cluster(this, 'myEventHubCluster', {
   *   resourceGroup: resourceGroup,
   *   name: 'myCluster',
   *   skuName: 'Dedicated_1', // This is optional since it defaults to 'Dedicated_1'
   *   tags: {
   *     department: 'IT'
   *   }
   * });
   * ```
   */
  constructor(scope: Construct, name: string, ehClusterProps: ClusterProps) {
    super(scope, name);

    this.ehClusterProps = ehClusterProps;
    this.resourceGroup = ehClusterProps.resourceGroup;

    const defaults = {
      skuName: ehClusterProps.skuName || "Dedicated_1",
      tags: ehClusterProps.tags || {},
    };

    const ehCluster = new EventhubCluster(this, "ehcluster", {
      name: ehClusterProps.name,
      resourceGroupName: ehClusterProps.resourceGroup.name,
      location: ehClusterProps.resourceGroup.location,
      ...defaults,
    });

    this.id = ehCluster.id;

    // Outputs
    const cdktfTerraformOutputEventhubId = new cdktf.TerraformOutput(
      this,
      "id",
      {
        value: ehCluster.id,
      },
    );
    cdktfTerraformOutputEventhubId.overrideLogicalId("id");
  }
}
