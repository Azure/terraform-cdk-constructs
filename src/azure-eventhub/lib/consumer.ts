import { EventhubConsumerGroup } from "@cdktf/provider-azurerm/lib/eventhub-consumer-group";
import { TerraformOutput } from "cdktf";
import { Construct } from "constructs";



export interface ConsumerGroupProps {
  readonly name: string;
  /**
   * The name of the resource group in which the EventHub Consumer Group's grandparent Namespace exists.
   */
  readonly resourceGroupName: string;
  /**
   * Specifies the name of the grandparent EventHub Namespace.
   */
  readonly namespaceName: string;
  /**
   * Specifies the name of the EventHub.
   */
  readonly eventhubName: string;
  /**
   * Specifies the user metadata.
   */
  readonly userMetadata?: string;
}

export class ConsumerGroup extends Construct {
  readonly ehConsumerGroupProps: ConsumerGroupProps;
  readonly id: string;

  constructor(scope: Construct, name: string, ehConsumerGroupProps: ConsumerGroupProps) {
    super(scope, name);

    this.ehConsumerGroupProps = ehConsumerGroupProps;

    const eventhubConsumerGroup = new EventhubConsumerGroup(this, ehConsumerGroupProps.name, {
      name: ehConsumerGroupProps.name,
      resourceGroupName: ehConsumerGroupProps.resourceGroupName,
      namespaceName: ehConsumerGroupProps.namespaceName,
      eventhubName: ehConsumerGroupProps.eventhubName,
      userMetadata: ehConsumerGroupProps.userMetadata,
    })

    this.id = eventhubConsumerGroup.id;

    // Outputs
    const cdktfTerraformOutputEventhubConsumerGroupId = new TerraformOutput(this, 'id', {
      value: eventhubConsumerGroup.id,
    });
    cdktfTerraformOutputEventhubConsumerGroupId.overrideLogicalId('id');
  }
}