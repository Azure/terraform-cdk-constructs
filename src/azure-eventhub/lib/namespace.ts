import { EventhubNamespace } from "@cdktf/provider-azurerm/lib/eventhub-namespace";
import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import { Instance, BaseInstanceProps } from "./instance";
import { AzureResourceWithAlert } from "../../core-azure/lib";

export interface NamespaceProps {
  /**
   * The Azure Resource Group in which to create the EventHub Namespace.
   */
  readonly resourceGroup: ResourceGroup;
  /**
   * The name of the EventHub Namespace to create.
   */
  readonly name: string;
  /**
   * Defines which tier to use. Valid options are Basic, Standard, and Premium.
   * @default "Basic"
   */
  readonly sku?: string;
  /**
   * Specifies the Capacity / Throughput Units for a Standard SKU namespace.
   * @default 2
   */
  readonly capacity?: number;
  /**
   * Specifies if the EventHub Namespace should be Auto Inflate enabled.
   * @default false
   */
  readonly autoInflateEnabled?: boolean;
  /**
   * Specifies the maximum number of throughput units when Auto Inflate is Enabled. Valid values range from 1 - 20.
   * @default 2
   */
  readonly maximumThroughputUnits?: number;
  /**
   * Specifies if the EventHub Namespace should be Zone Redundant (created across Availability Zones).
   * @default true
   */
  readonly zoneRedundant?: boolean;
  /**
   * The tags to assign to the Key Vault.
   */
  readonly tags?: { [key: string]: string };
  /**
   * The minimum supported TLS version for this EventHub Namespace. Valid values are: 1.0, 1.1 and 1.2.
   * @default "1.2"
   */
  readonly minimumTlsVersion?: string;
  /**
   * Is public network access enabled for the EventHub Namespace?
   * @default true
   */
  readonly publicNetworkAccessEnabled?: boolean;
  /**
   * Is SAS authentication enabled for the EventHub Namespace? North Central US Not supported.
   * @default false
   */
  readonly localAuthenticationEnabled?: boolean;
  /**
   * Specifies the type of Managed Service Identity that should be configured on this Event Hub Namespace. Possible values are SystemAssigned or UserAssigned.
   * @default "SystemAssigned"
   */
  readonly identityType?: string;
  /**
   * Specifies a list of User Assigned Managed Identity IDs to be assigned to this EventHub namespace.
   */
  readonly identityIds?: string[] | undefined;

  /**
   * TODO: network_rulesets
   */
}

export class Namespace extends AzureResourceWithAlert {
  readonly ehNamespaceProps: NamespaceProps;
  public resourceGroup: ResourceGroup;
  public id: string;
  readonly namespaceName: string;

  /**
   * Constructs a new Event Hub Namespace.
   *
   * This class creates an Azure Event Hub Namespace, which serves as a container for all messaging components.
   * Namespaces provide a scoping container for Event Hub resources within a specific region, which can be further
   * controlled and managed using the provided settings.
   *
   * @param scope - The scope in which to define this construct, typically representing the Cloud Development Kit (CDK) stack.
   * @param name - The unique name for this instance of the Event Hub Namespace.
   * @param ehNamespaceProps - The properties for configuring the Event Hub Namespace. These properties include:
   *                - `resourceGroup`: Required. The Azure Resource Group in which the namespace will be created.
   *                - `name`: Required. The name of the Event Hub Namespace to create.
   *                - `sku`: Optional. The SKU tier of the namespace (Basic, Standard, Premium). Defaults to "Basic".
   *                - `capacity`: Optional. Specifies the throughput units for a Standard SKU namespace. Defaults to 2.
   *                - `autoInflateEnabled`: Optional. Enables or disables Auto Inflate. Defaults to false.
   *                - `maximumThroughputUnits`: Optional. The maximum number of throughput units when Auto Inflate is enabled. Defaults to 2.
   *                - `zoneRedundant`: Optional. Specifies if the namespace should be zone redundant. Defaults to true.
   *                - `tags`: Optional. Tags for resource management and categorization.
   *                - `minimumTlsVersion`: Optional. Specifies the minimum supported TLS version. Defaults to "1.2".
   *                - `publicNetworkAccessEnabled`: Optional. Specifies if public network access is enabled. Defaults to true.
   *                - `localAuthenticationEnabled`: Optional. Specifies if SAS authentication is enabled. Defaults to false.
   *                - `identityType`: Optional. The type of Managed Service Identity. Defaults to "SystemAssigned".
   *                - `identityIds`: Optional. A list of User Assigned Managed Identity IDs.
   *
   * Example usage:
   * ```typescript
   * const eventHubNamespace = new Namespace(this, 'myNamespace', {
   *   resourceGroup: resourceGroup,
   *   name: 'myEventHubNamespace',
   *   sku: 'Standard',
   *   capacity: 4,
   *   autoInflateEnabled: true,
   *   maximumThroughputUnits: 10,
   *   zoneRedundant: false,
   *   tags: {
   *     department: 'IT'
   *   },
   *   minimumTlsVersion: '1.2',
   *   publicNetworkAccessEnabled: false,
   *   localAuthenticationEnabled: true,
   *   identityType: 'SystemAssigned'
   * });
   * ```
   */
  constructor(
    scope: Construct,
    name: string,
    ehNamespaceProps: NamespaceProps,
  ) {
    super(scope, name);

    this.ehNamespaceProps = ehNamespaceProps;
    this.resourceGroup = ehNamespaceProps.resourceGroup;
    this.namespaceName = ehNamespaceProps.name;

    const defaults = {
      sku: ehNamespaceProps.sku || "Basic",
      capacity: ehNamespaceProps.capacity || 2,
      autoInflateEnabled: ehNamespaceProps.autoInflateEnabled || false,
      maximumThroughputUnits: ehNamespaceProps.maximumThroughputUnits || 2,
      zoneRedundant: ehNamespaceProps.zoneRedundant || false,
      tags: ehNamespaceProps.tags || {},
      minimumTlsVersion: ehNamespaceProps.minimumTlsVersion || "1.2",
      publicNetworkAccessEnabled:
        ehNamespaceProps.publicNetworkAccessEnabled || true,
      localAuthenticationEnabled:
        ehNamespaceProps.localAuthenticationEnabled || true,
      identity: {
        type: ehNamespaceProps.identityType || "SystemAssigned",
        identityIds:
          ehNamespaceProps.identityType == "UserAssigned"
            ? ehNamespaceProps.identityIds
            : undefined,
      },
    };

    const eventhubNamespce = new EventhubNamespace(this, "ehnamespace", {
      name: ehNamespaceProps.name,
      resourceGroupName: this.resourceGroup.name,
      location: this.resourceGroup.location,
      ...defaults,
    });

    // Outputs
    this.id = eventhubNamespce.id;
    const cdktfTerraformOutputEventhubNamespceId = new cdktf.TerraformOutput(
      this,
      "id",
      {
        value: eventhubNamespce.id,
      },
    );
    cdktfTerraformOutputEventhubNamespceId.overrideLogicalId("id");
  }

  /**
   * Creates and adds an Event Hub instance to the current namespace.
   *
   * This method sets up a new Event Hub instance within the namespace defined by this class. An Event Hub instance
   * serves as a container that processes and stores events. This method facilitates the setup of multiple Event Hubs
   * within a single namespace, each configured according to the specified properties.
   *
   * @param props - The properties for configuring the new Event Hub instance. These properties extend `BaseInstanceProps`, which include:
   *                - `name`: Required. The name of the Event Hub instance.
   *                - `partitionCount`: Optional. The number of partitions in the Event Hub. Default is 2.
   *                - `messageRetention`: Optional. The number of days to retain messages in the Event Hub. Default is 1.
   *                - `status`: Optional. The operational status of the Event Hub (Active, Disabled, SendDisabled). Default is "Active".
   *                Other properties from `BaseInstanceProps` can also be passed and will be used in the creation of the Event Hub.
   *
   * @returns An instance of the Event Hub (`Instance` class), configured with the specified properties.
   *
   * Example usage:
   * ```typescript
   * const eventHub = namespace.addEventhubInstance({
   *   name: 'myEventHub',
   *   partitionCount: 4,
   *   messageRetention: 7,
   *   status: 'Active'
   * });
   * ```
   *
   * @remarks
   * Ensure that the namespace has sufficient capacity and configuration to support the properties of the Event Hub being created,
   * especially in terms of partition count and throughput units if applicable.
   */
  addEventhubInstance(props: BaseInstanceProps) {
    return new Instance(this, "ehinstance", {
      resourceGroup: this.resourceGroup,
      namespaceName: this.namespaceName,
      ...props,
    });
  }
}
