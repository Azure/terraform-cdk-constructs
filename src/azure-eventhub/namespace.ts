import { EventhubNamespace } from "@cdktf/provider-azurerm/lib/eventhub-namespace";
import { Construct } from "constructs";
import * as cdktf from 'cdktf';
import { AzureResourceGroup } from "../azure-resourcegroup";
import { AzureEventhubInstance, AzureEventhubInstanceProps } from './eventhub';
import { AzureResource } from "../core-azure";


export interface AzureEventhubNamespaceProps {
	/**
	 * The Azure Resource Group in which to create the EventHub Namespace.
	 */
	readonly rg: AzureResourceGroup
	/**
	 * The name of the EventHub Namespace to create.
	 */
	readonly name: string;
	/**
	* Defines which tier to use. Valid options are Basic, Standard, and Premium.
	* @default "Basic"
	*/
	readonly sku?: string;
	/**
	 * Specifies the Capacity / Throughput Units for a Standard SKU namespace. 
	 * @default 2
	 */
	readonly capacity?: number;
	/**
	 * Specifies if the EventHub Namespace should be Auto Inflate enabled.
	 * @default false
	 */
	readonly autoInflateEnabled?: boolean;
	/**
	 * Specifies the maximum number of throughput units when Auto Inflate is Enabled. Valid values range from 1 - 20.
		* @default 2
	 */
	readonly maximumThroughputUnits?: number;
	/**
	 * Specifies if the EventHub Namespace should be Zone Redundant (created across Availability Zones).
	 * @default true
	 */
	readonly zoneRedundant?: boolean;
	/**
		 * The tags to assign to the Key Vault.
		 */
	readonly tags?: { [key: string]: string; };
	/**
	 * The minimum supported TLS version for this EventHub Namespace. Valid values are: 1.0, 1.1 and 1.2.
	 * @default "1.2"
	 */
	readonly minimumTlsVersion?: string;
	/**
	 * Is public network access enabled for the EventHub Namespace?
	 * @default true
	 */
	readonly publicNetworkAccessEnabled?: boolean;
	/**
	 * Is SAS authentication enabled for the EventHub Namespace? North Central US Not supported.
	 * @default false
	 */
	readonly localAuthenticationEnabled?: boolean;
	/**
	 * Specifies the type of Managed Service Identity that should be configured on this Event Hub Namespace. Possible values are SystemAssigned or UserAssigned.
	 * @default "SystemAssigned"
	 */
	readonly identityType?: string;
	/**
	 * Specifies a list of User Assigned Managed Identity IDs to be assigned to this EventHub namespace.
	 */
	readonly identityIds?: string[] | undefined;

	/**
	 * TODO: network_rulesets
	 */
}

export class AzureEventhubNamespace extends AzureResource {
	readonly ehNamespaceProps: AzureEventhubNamespaceProps;
	readonly rgName: string;
	readonly rgLocation: string;
	readonly id: string;
	readonly namespaceName: string;

	constructor(scope: Construct, name: string, ehNamespaceProps: AzureEventhubNamespaceProps) {
		super(scope, name);

		this.ehNamespaceProps = ehNamespaceProps;
		this.rgName = ehNamespaceProps.rg.Name;
		this.rgLocation = ehNamespaceProps.rg.Location;
		this.namespaceName = ehNamespaceProps.name;

		const defaults = {
			sku: ehNamespaceProps.sku || 'Basic',
			capacity: ehNamespaceProps.capacity || 2,
			autoInflateEnabled: ehNamespaceProps.autoInflateEnabled || false,
			maximumThroughputUnits: ehNamespaceProps.maximumThroughputUnits || 2,
			zoneRedundant: ehNamespaceProps.zoneRedundant || false,
			tags: ehNamespaceProps.tags || {},
			minimumTlsVersion: ehNamespaceProps.minimumTlsVersion || '1.2',
			publicNetworkAccessEnabled: ehNamespaceProps.publicNetworkAccessEnabled || true,
			localAuthenticationEnabled: ehNamespaceProps.localAuthenticationEnabled || true,
			identity: {
				type: ehNamespaceProps.identityType || 'SystemAssigned',
				identityIds: ehNamespaceProps.identityType == 'UserAssigned' ? ehNamespaceProps.identityIds : undefined,
			}
		}

		const eventhubNamespce = new EventhubNamespace(this, `ehnamespace`, {
			name: ehNamespaceProps.name,
			resourceGroupName: this.rgName,
			location: this.rgLocation,
			...defaults,
		})

		// Outputs
		this.id = eventhubNamespce.id;
		const cdktfTerraformOutputEventhubNamespceId = new cdktf.TerraformOutput(this, 'id', {
			value: eventhubNamespce.id,
		});
		cdktfTerraformOutputEventhubNamespceId.overrideLogicalId('id');
	}

	// Create an EventHub Instance Method
	addEventhubInstance(props: Omit<AzureEventhubInstanceProps, 'resourceGroupName' | 'namespaceName'>) {
		return new AzureEventhubInstance(this, `ehinstance`, {
			resourceGroupName: this.rgName,
			namespaceName: this.namespaceName,
			...props,
		});
	}
}