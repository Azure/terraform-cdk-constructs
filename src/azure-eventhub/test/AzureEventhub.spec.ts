import { Testing, TerraformStack} from 'cdktf';
import 'cdktf/lib/testing/adapters/jest';
import {AzurermProvider} from "@cdktf/provider-azurerm/lib/provider";
import * as util from "../../util/azureTenantIdHelpers";
import { exampleAzureEventhub } from './ExampleAzureEventhub';
import { AzureEventhubNamespace } from '../namespace';

describe('Azure Eventhub With Defaults', () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    jest.mock('../../azure-resourcegroup', () => ({
      AzureResourceGroup: {
        Location: 'eastus',
        Name: `test-rg`,
      }
    }));
    const rgMock = jest.requireMock('../../azure-resourcegroup');

    const app = Testing.app();
    stack = new TerraformStack(app, 'test');

    new AzurermProvider(stack, "azureFeature", {features: {}});
    new AzureEventhubNamespace(stack, 'testAzureEventhubDefaults', rgMock.AzureResourceGroup, {
      name: `eh-test`,
      tags: {
        'test': 'test'
      }
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Azure Eventhub with defaults and checks snapshot", () => {
    expect(
      Testing.synth(stack)
    ).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});


describe('Azure Eventhub Example', () => {

  it("renders the Example Azure Eventhub and checks snapshot", () => {
    // Need to remove the tenant_id from the snapshot as it will change wherever the test is run
    const output = Testing.synth(new exampleAzureEventhub(Testing.app(), "testAzureEventhub"));
    const myObject: Record<string, any> = JSON.parse(output);

    expect(util.removeTenantIdFromSnapshot(myObject)).toMatchSnapshot();
});

it("check if the produced terraform configuration is valid", () => {
  // We need to do a full synth to plan the terraform configuration
  expect(Testing.fullSynth(new exampleAzureEventhub(Testing.app(), "testAzureEventhub"))).toBeValidTerraform();
});

it("check if this can be planned", () => {

  
  // We need to do a full synth to plan the terraform configuration
  expect(Testing.fullSynth(new exampleAzureEventhub(Testing.app(), "testAzureEventhub"))).toPlanSuccessfully();
});
});
