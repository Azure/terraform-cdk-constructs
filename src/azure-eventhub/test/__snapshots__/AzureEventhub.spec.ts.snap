// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Azure Eventhub Example renders the Example Azure Eventhub and checks snapshot 1`] = `
{
  "data": {
    "azurerm_client_config": {
      "CurrentClientConfig": {},
    },
  },
  "output": {
    "id": {
      "value": "\${azurerm_monitor_metric_alert.eventhub_metricalert_metricAlert_98164A67.id}",
    },
    "partition_ids": {
      "value": "\${tolist(azurerm_eventhub.eventhub_ehinstance_ehinstance-test-eventhub-instance_6C544BA5.partition_ids)}",
    },
    "primary_connection_string": {
      "sensitive": true,
      "value": "\${azurerm_eventhub_authorization_rule.eventhub_ehinstance_ehauthrule-test-rule_80C99DBB.primary_connection_string}",
    },
    "primary_key": {
      "sensitive": true,
      "value": "\${azurerm_eventhub_authorization_rule.eventhub_ehinstance_ehauthrule-test-rule_80C99DBB.primary_key}",
    },
  },
  "provider": {
    "azurerm": [
      {
        "features": {},
      },
    ],
  },
  "resource": {
    "azurerm_eventhub": {
      "eventhub_ehinstance_ehinstance-test-eventhub-instance_6C544BA5": {
        "message_retention": 1,
        "name": "test-eventhub-instance",
        "namespace_name": "ehns-\${var.name}",
        "partition_count": 2,
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "status": "Active",
      },
    },
    "azurerm_eventhub_authorization_rule": {
      "eventhub_ehinstance_ehauthrule-test-rule_80C99DBB": {
        "eventhub_name": "test-eventhub-instance",
        "listen": true,
        "manage": false,
        "name": "test-rule",
        "namespace_name": "ehns-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "send": true,
      },
    },
    "azurerm_eventhub_consumer_group": {
      "eventhub_ehinstance_ehconsumergroup-test-consumer-group_3A1228DF": {
        "eventhub_name": "test-eventhub-instance",
        "name": "test-consumer-group",
        "namespace_name": "ehns-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
      },
    },
    "azurerm_eventhub_namespace": {
      "eventhub_ehnamespace_921C1DD7": {
        "auto_inflate_enabled": false,
        "capacity": 2,
        "identity": {
          "type": "SystemAssigned",
        },
        "local_authentication_enabled": true,
        "location": "\${azurerm_resource_group.rg.location}",
        "maximum_throughput_units": 2,
        "minimum_tls_version": "1.2",
        "name": "ehns-\${var.name}",
        "public_network_access_enabled": true,
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "sku": "Basic",
        "tags": {},
        "zone_redundant": false,
      },
    },
    "azurerm_kusto_eventhub_data_connection": {
      "eventhub_ehinstance_ehkustodataconnection-test-eventhub-instance-test-kusto-data-connection_kusto-data-connection-test-kusto-data-connection_9E92DF1E": {
        "cluster_name": "testkustocluster",
        "compression": "None",
        "consumer_group": "$Default",
        "data_format": "JSON",
        "database_name": "test-kusto-database",
        "database_routing_type": "Single",
        "eventhub_id": "\${azurerm_eventhub.eventhub_ehinstance_ehinstance-test-eventhub-instance_6C544BA5.id}",
        "location": "eastus",
        "name": "test-kusto-data-connection",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
      },
    },
    "azurerm_monitor_metric_alert": {
      "eventhub_metricalert_metricAlert_98164A67": {
        "auto_mitigate": true,
        "criteria": [
          {
            "aggregation": "Total",
            "metric_name": "Server Errors.",
            "metric_namespace": "Microsoft.EventHub/namespaces",
            "operator": "GreaterThan",
            "skip_metric_validation": false,
            "threshold": 100,
          },
        ],
        "enabled": true,
        "frequency": "PT5M",
        "name": "test-metric-alert",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "scopes": [
          "\${azurerm_eventhub_namespace.eventhub_ehnamespace_921C1DD7.id}",
        ],
        "severity": 3,
        "window_size": "PT5M",
      },
    },
    "azurerm_resource_group": {
      "rg": {
        "location": "eastus",
        "name": "rg-\${var.name}",
      },
    },
    "azurerm_role_assignment": {
      "eventhub_TfTokenTOKEN57Contributor_role_26F7B48B": {
        "principal_id": "\${data.azurerm_client_config.CurrentClientConfig.object_id}",
        "role_definition_name": "Contributor",
        "scope": "\${azurerm_eventhub_namespace.eventhub_ehnamespace_921C1DD7.id}",
      },
    },
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0",
      },
    },
  },
  "variable": {
    "name": {
      "default": "test",
      "description": "System name used to randomize the resources",
      "type": "string",
    },
  },
}
`;

exports[`Azure Eventhub With Defaults renders an Azure Eventhub with defaults and checks snapshot 1`] = `
"{
  "output": {
    "id": {
      "value": "\${azurerm_eventhub_namespace.testAzureEventhubDefaults_ehnamespace_B16BC9EF.id}"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_eventhub_namespace": {
      "testAzureEventhubDefaults_ehnamespace_B16BC9EF": {
        "auto_inflate_enabled": false,
        "capacity": 2,
        "identity": {
          "type": "SystemAssigned"
        },
        "local_authentication_enabled": true,
        "location": "\${azurerm_resource_group.MyResourceGroup.location}",
        "maximum_throughput_units": 2,
        "minimum_tls_version": "1.2",
        "name": "eh-test",
        "public_network_access_enabled": true,
        "resource_group_name": "\${azurerm_resource_group.MyResourceGroup.name}",
        "sku": "Basic",
        "tags": {
          "test": "test"
        },
        "zone_redundant": false
      }
    },
    "azurerm_resource_group": {
      "MyResourceGroup": {
        "location": "eastus",
        "name": "rg-test"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0"
      }
    }
  }
}"
`;
